contracts/lotterybook/merkletree/merkletree.go

```diff
@@ -0,0 +1,393 @@
+// by their weight bias.
+type EntryByBias []*Entry
+
+func (s EntryByBias) Len() int           { return len(s) }
+func (s EntryByBias) Less(i, j int) bool { return s[i].bias < s[j].bias }
+func (s EntryByBias) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// EntryByLevel implements the sort interface to allow sorting a list of entries
+// by their position in the tree in descending order.
+type EntryByLevel []*Entry
+
+func (s EntryByLevel) Len() int           { return len(s) }
+func (s EntryByLevel) Less(i, j int) bool { return s[i].level > s[j].level }
+func (s EntryByLevel) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+// Node represents a node in merkle tree.
+type Node struct {
+	Nodehash common.Hash // The hash of node.
+	Parent   *Node       // The parent of this node, nil if it's root node.
+	Left     *Node       // The left child of this node
+	Right    *Node       // The right child of this node
+	Level    uint64      // The level of node in this tree
+	Value    *Entry      // The referenced entry by this node, nil if it's not leaf.
+}
+
+// Hash returns the hash of this tree node.
+func (node *Node) Hash() common.Hash {
+	// Short circuit if nodehash is already cached.
+	if node.Nodehash != (common.Hash{}) {
+		return node.Nodehash
+	}
+	// If it's a leaf node, derive the hash by the entry content.
+	if node.Value != nil {
+		node.Nodehash = node.Value.Hash()
+		return node.Nodehash
+	}
+	// It's a branch node, derive the hash via two children.
+	left, right := node.Left.Hash(), node.Right.Hash() // Both children should never be nil.
+	if bytes.Compare(left.Bytes(), right.Bytes()) < 0 {
+		node.Nodehash = crypto.Keccak256Hash(append(left.Bytes(), right.Bytes()...))
+	} else {
+		node.Nodehash = crypto.Keccak256Hash(append(right.Bytes(), left.Bytes()...))
+	}
+	return node.Nodehash
+}
+
+// String returns the string format of node.
+func (node *Node) String() string {
+	if node.Value != nil {
+		return fmt.Sprintf("E(%x:%d)", node.Value.Value, node.Value.level)
+	}
+	return fmt.Sprintf("N(%x) => L.(%s) R.(%s)", node.Hash(), node.Left.String(), node.Right.String())
+}
+
+type MerkleTree struct {
+	Roothash common.Hash // The hash of root node, maybe null if we never calculate it.
+	Root     *Node       // The root node of merkle tree.
+	Leaves   []*Node     // Batch of leaves node included in the tree.
+}
+
+// NewMerkleTree constructs a merkle tree with given entries.
+func NewMerkleTree(entries []*Entry) (*MerkleTree, map[string]struct{}, error) {
+	if len(entries) == 0 {
+		return nil, nil, ErrEmptyEntryList
+	}
+	// Assign an unique salt for each entry. The hash
+	// of entry is calculated by keccak256(value, salt)
+	// so we can preserve the privacy of given value.
+	for _, entry := range entries {
+		entry.salt = rand.Uint64()
+	}
+	// Verify the validity of the given entries.
+	var sum, totalWeight uint64
+	for _, entry := range entries {
+		if entry.Weight == 0 {
+			return nil, nil, ErrInvalidWeight
```

Comment:
Since you allow dropping small entries now, these can be dropped too without an error. If `newTree` is made error free (see my comment there) then this function can also never return an error which would be really nice.

Reply:
Ok. I can change it. 

Originally I think it's quite easy to avoid this error because in theory we never ever want to pass the entry with zero weight. 

But it's fine to add them to dropped map.